Makefile

```diff
@@ -1,7 +1,28 @@
-.PHONY: lint
+VENV=venv
+PIP_LOG=pip-log.txt
+PRECOMMIT_SRC=.githooks/pre-commit
+PRECOMMIT_DST=.git/hooks/pre-commit
+
+export VIRTUAL_ENV := $(abspath ${VENV})
+export PATH := ${VIRTUAL_ENV}/bin:${PATH}
+
+${PRECOMMIT_DST}: ${PRECOMMIT_SRC}
+	cp $< $@
+
+${PIP_LOG}: requirements.txt requirements-dev.txt
+	if [[ ! -e ${VENV}/bin/activate ]]; then virtualenv --python=python3 ${VENV}; fi
```

Comment:
If the requirement files have changed you'd want to recreate the virtualenv. I'd rather just
```
${VENV}: requirements.txt requirements-dev.txt
    rm -rf ${VENV}
    virtualenv --python=python3 ${VENV}
    {VENV}/bin/pip install -r requirements.txt -r requirements-dev.txt
```

We usually https://tox.readthedocs.io/en/latest/ to do this automatically,. But since most zipkin people don't have standard python tools installed, recreating the virtualenv seems easier.

Reply:
Yep, `tox` would be the clean solution to this, but as you point out, it'll pay to keep this as lean as possible.

I was going to argue that I want to minimize the time it takes to run `make`, but this will only run anyway when the dependencies change, which I expect to be extremely rare. Might as well rebuild the virtualenv in those cases.

---

We're still missing out on installing upgrades to packages released since the venv was created, but that's the price we way for not using lock files and/or `tox`. The Docker stuff doesn't solve this either, since we're often using the build cache. Maybe it'd be better to `ADD` all the source files before invoking `pip install`, making sure we always run with dependencies at least as new as the last code change... which might still not be very new in the long run. So some TTL-based stuff might be needed there, or just plain running with the build cache disabled, paying the extra setup time on each run. I really want to avoid that.

Reply:
The reason why we usually recreate the virtualenv is that if you remove a dependency from requirements.txt, `pip install` won't remove it. Which might led to the script being broken for other people if that dependency was actually still required.

Reply:
FWIW, the Docker approach *does* take care of that. Regardless, I'm convinced rebuilding is the right choice here. Will get around to it in a couple of days.